torch/__init__.py

```diff
@@ -426,11 +426,41 @@ def use_deterministic_algorithms(d):
     Note that deterministic operations tend to have worse performance than
     non-deterministic operations.
 
+    .. note::
+
+        :func:`use_deterministic_algorithms` is not guaranteed to prevent
+        nondeterministic behavior that is caused by an operation performing
+        multiple writes of different data to a single memory location. The
+        order of writes to a single location is not guaranteed, which can cause
+        nondeterminism. This could occur when an output tensor contains internal
+        memory overlaps or if an indexed operation is given duplicate indices.
+        Below are several examples.
+
+    .. code-block:: python
+
+        >>> torch.use_deterministic_algorithms(True)
+
+        # Determinism is not guaranteed for an in-place operation on a tensor
```

Comment:
We might as well add the examples in the revision of this PR (if they're a pain we can always separate them to another PR). Example ideas from our notes:

- forward
- backward
- overlapped tensor
- writing to multiple locations (which will throw a nondeterminism error)

Reply:
By "writing to multiple locations", I assume you're talking about duplicate indices. Since they will throw nondeterminism errors, I don't think we need an example anymore.

As mentioned above (https://github.com/pytorch/pytorch/pull/55413#discussion_r609071768), it will be difficult to maintain a working example for overlapped tensors as time goes on. Perhaps we shouldn't include any examples then? Or at least we could think about it an add them in another PR.

Reply:
Yes, sorry, "writing to multiple locations" was the opposite of what I meant. I edited my comment for clarity.

Removing the overlapped example sounds good.